bin/post_process_sce.R

```diff
@@ -97,13 +149,53 @@ filtered_sce <- filtered_sce |>
 # replace existing stats from altExp if any
 for (alt in altExpNames(filtered_sce)) {
   # remove old row data
-  drop_cols = colnames(rowData(altExp(filtered_sce, alt))) %in% c('mean', 'detected')
+  drop_cols <- colnames(rowData(altExp(filtered_sce, alt))) %in% c('mean', 'detected')
   rowData(altExp(filtered_sce, alt)) <- rowData(altExp(filtered_sce, alt))[!drop_cols]
 
   # add alt experiment features stats for filtered data
   altExp(filtered_sce, alt) <- scuttle::addPerFeatureQCMetrics(altExp(filtered_sce, alt))
 }
 
+# filter sce based on ADTs
+if (process_adt) {
+
+  adt_qc_df <- DropletUtils::cleanTagCounts(
+    counts(altExp(filtered_sce, altexp_name)),
+    ambient = ambient_profile
+  )
+
+  # Filter according to `discard`
+  filtered_sce <- filtered_sce[, which(!(adt_qc_df$discard))]
+
+  # TODO: Should we add `adt_qc_df` into the altExp? If we want it:
+  #colData(altExp(filtered_sce, altexp_name)) <- cbind(colData(altExp(filtered_sce, altexp_name)), adt_qc_df)
+
+  # Filter on negative controls, if present
+  neg_controls <- adt_barcode_df |>
+    dplyr::filter(target_type == "neg_control") |>
+    dplyr::pull(name)
+
+  if (length(neg_controls) > 0) {
+    # counts for each negative control
+    neg_control_counts <- counts(altExp(filtered_sce, altexp_name))[neg_controls, ]
+
+    # remove any cell with >=threshold for any given negative control
+    # first, determine which cells need to be removed
+    cells_to_remove <- purrr::map(
+      neg_controls,
+      \(x) which(neg_control_counts[x,] >= opt$adt_negative_cutoff)
+    ) |>
+    unlist() |>
+    unique()
```

Comment:
My general philosophy is that if we have a matrix, let's use the matrix:
    
```suggestion
    # remove any cell with >=threshold for any given negative control
    cells_to_remove <- colSums(neg_control_counts >= opt$adt_negative_cutoff) > 0
```
```

Reply:
for complete parallelism, it should be:
```
cells_to_remove <- which(colSums(neg_control_counts >= opt$adt_negative_cutoff) > 0)
```

Reply:
> No, my code is checking each cell, converting it to a boolean matrix, then adding the sums of TRUE and FALSE for each cell.

Ah, indeed! It would help if I more precisely noted where the parentheses are!

Reply:
Minor followup: possibly better to do: 
```
cells_to_keep <- which(colSums(neg_control_counts >= opt$adt_negative_cutoff) == 0)
```
and save a negation